README.md
# crawler-toolbox

爬虫工具箱

v0.2.0 2020-4-3

## 用途

* 作为 [HttpToolbox](https://github.com/ufologist/HttpToolbox) 的 Node.js 版
* 列举了爬虫抓取网站的一般场景
* 作为开发爬虫项目的模版

## 依赖

* 爬虫的核心模块: HTTP 与 解析 DOM
  * `superagent` HTTP
  * `cheerio` 解析 DOM
* 日志
  * `log4js`
* 命令行
  * `commander`
  * `inquirer`
  * `chalk`
* 定时任务
  * `node-schedule` cron 定时任务
* 工具
  * `sanitize-filename` 避免写文件时由于非法的文件名造成写入失败
  * `superagent-charset` 针对非 UTF-8 编码的网站
  * `beepbeep` 声音提醒
  * `node-notifier` 桌面提醒
  * 钉钉机器人提醒

## 参考

* [网络爬虫与数据库操作](https://github.com/leizongmin/book-crawler-mysql-cron/blob/master/book.md)
  * 发起一个 HTTP 请求来获取指定 URL 的内容
  * 使用 jQuery 的查询语法来操作网页元素，提取出需要的数据
  * 将数据储存到数据库中，以及从数据库中查询出这些数据
  * 建立一个简单的 Web 服务器来显示这些数据
  * 使用简单的方法来让一些程序在指定的时间自动执行
  * 让程序更稳定地运行
  * 对一些常见的字符编码互相转换
* [神箭手-云爬虫开发平台](http://www.shenjianshou.cn)

  神箭手是一个一站式云端通用爬虫开发平台。海量代理IP、JS自动渲染、防屏蔽、图片云托管、验证码识别.

  [神箭手云爬虫 爬取规则示例](https://github.com/ShenJianShou/crawler_samples)
* [八爪鱼采集器 - 网页数据采集器](http://www.bazhuayu.com/)
* [Finndy+云采集引擎](http://www.finndy.com/?action-engine)
* [Faker](https://github.com/cupools/faker "An experience spider to collect free proxy ip through extensible task")
  * [西刺免费](http://www.xicidaili.com/nn/)
  * [快代理](http://www.kuaidaili.com/free/inha/)

CHANGELOG.md
# CHANGELOG

* v0.2.0 2020-4-2

  * feat: 添加 `node-notifier` 依赖用于桌面提醒(notification)
  * feat: 添加钉钉机器人用于移动端提醒

* v0.1.0 2016-12-7
  
  初始版本

crawler.js
// 系统模块
var fs = require('fs');

// HTTP 模块与 DOM 解析, 这是爬虫的核心模块
var request = require('superagent');
var charset = require('superagent-charset');
charset(request);

// 日志
var log4js = require('log4js');
log4js.configure({
    appenders: [{
        type: 'console'
    }, {
        type: 'logLevelFilter',
        level: 'warn',
        appender: {
            type: 'dateFile',
            filename: '_runtime_crawler',
            pattern: '-yyyy-MM-dd.log',
            alwaysIncludePattern: true
        }
    }],
    replaceConsole: true
});
var sanitize = require('sanitize-filename');

var crawlerUtil = require('./crawler-util.js');

// 如果一个抛出的异常没有被 try {} catch (err) {} 捕捉到,
// 其会尝试将这些错误交由 uncaughtException 事件处理程序来处理,
// 仅当没有注册该事件处理程序时, 才会最终导致进程直接退出
//
// 有时候, 由于 Node.js 自身的 Bug 或者使用到的第三方 C++ 模块的缺陷而导致一些底层的错误,
// 比如在 Linux 系统下偶尔会发生段错误(segment fault)导致进程崩溃,
// 此时上面提到的处理 uncaughtException 事件的方法就不适用了.
// pm2 是一个功能强大的进程管理器
process.on('uncaughtException', function (err) {
    console.error('uncaughtException: %s', err.stack);
});

// 简单的就直接使用 process.argv
// 复杂的就参考 crawler-example.js 中的 commander + inquirer
var args = process.argv.slice(2);

var userAgent = 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.104 Safari/537.36 Core/1.53.1708.400 QQBrowser/9.5.9635.400';
// 请求超时时间为 1 个小时
var timeout = 60 * 60 * 1000;

/**
 * 在请求发出去之前, 做一些通用的默认设置
 * 例如设置统一的超时时间, 统一的 User-Agent 头
 */
function buildAgent(agent) {
    agent.set('User-Agent', userAgent)
        //  .set('X-Requested-With', 'XMLHttpRequest') // ajax 请求一般会添加这个头
         .timeout(timeout);

    return agent;
}

// HTTP 请求相关的示例请参考 crawler-example.js
function demo(keyword) {
    // 抓取百度搜索的结果
    var agent = request.get('https://www.baidu.com/s');
    buildAgent(agent);

    return agent.query({
                    wd: keyword
                })
                .then(crawlerUtil.showMeTheDollar)
                .then(function($) {
                    return $('.result').map(function() {
                        var $this = $(this);

                        return {
                            title: $this.find('.t').text(),
                            url: $this.find('.c-showurl').text()
                        };
                    }).toArray();
                });
}

demo('亚马逊').then(function(result) {console.log(result);});

crawler-util.js
var crypto = require('crypto');

// 解析 DOM
var cheerio = require('cheerio');
var request = require('superagent');

/**
 * show me the money
 * 解析 HTML, 返回一个类似 jQuery 的对来用于解析 DOM 结构
 */
function showMeTheDollar(response) {
    var $ = cheerio.load(response.text, {
        // 中文乱码？不，是 HTML 实体编码！
        // http://www.cnblogs.com/zichi/p/5135636.html
        decodeEntities: false
    });

    return $;
}

// npm install chance
// chance.integer({min: 1, max: 2});
function randomInt(min, max) {
    // https://github.com/sindresorhus/random-int
    return Math.floor(Math.random() * (max - min + 1) + min);
}

/**
 * 钉钉机器人加签
 * 
 * @return {string}
 */
function dingdingRobotSign(timestamp, secret) {
    var plainText = timestamp + '\n' + secret;
    var cipherText = crypto.createHmac('sha256', secret)
                           .update(plainText)
                           .digest()
                           .toString('base64');
    return cipherText;
}

/**
 * 推送钉钉机器人消息
 * 
 * @param {string} webhook 
 * @param {string} secret 
 * @param {string | object} message 
 * @param {function} [buildAgent] 
 * 
 * @see https://ding-doc.dingtalk.com/doc#/serverapi2/qf2nxq/e9d991e2
 */
function sendDingdingRobotMessage(webhook, secret, message, buildAgent) {
    var timestamp = Date.now();
    var sign = dingdingRobotSign(timestamp, secret);
    var _webhook = '';
    var _message = typeof message === 'string' ? {
        msgtype: 'text', 
        text: {
            content: message
        }
    } : message;

    if (webhook.indexOf('?') === -1) {
        _webhook = `${webhook}?timestamp=${timestamp}&sign=${encodeURIComponent(sign)}`;
    } else {
        _webhook = `${webhook}&timestamp=${timestamp}&sign=${encodeURIComponent(sign)}`;
    }

    var agent = request.post(_webhook);
    buildAgent && buildAgent(agent);

    return agent.send(_message)
                .then(function(result) {
                    return result.body;
                });
}

module.exports = {
    showMeTheDollar: showMeTheDollar,
    randomInt: randomInt,
    sendDingdingRobotMessage: sendDingdingRobotMessage
};

crawler-example.js
// 系统模块
var fs = require('fs');

var dayjs = require('dayjs');

// HTTP 模块与 DOM 解析, 这是爬虫的核心模块
// HTTP 请求
var request = require('superagent'); // 备选模块 request, 也支持 cookie
// node爬虫之gbk网页中文乱码解决方案
// http://www.cnblogs.com/zichi/p/5157887.html
var charset = require('superagent-charset');
charset(request);
// ENOENT thrown when trying to create a file with an invalid filename
// https://github.com/nodejs/node/issues/8987
var sanitize = require('sanitize-filename');

// encodeURIComponent with charset
// https://github.com/alsotang/urlencode

// 辅组类模块
// 解析命令行参数
var commander = require('commander');
// 命令行输入
var inquirer = require('inquirer');
// 命令行输出的颜色
var chalk = require('chalk');
// 如果有比较复杂的输入选择项, 则使用 commander + inquirer 来处理
commander.version('0.1.0')
         .description('我是一个命令行工具')
         .option('-a, --aa [参数1]', '参数1')
         .option('-b, --bb [参数2]', '参数2')
         .parse(process.argv);
console.log('commander', commander.aa, commander.bb);
// 如果没有传入参数, 则直接提示帮助信息后退出
// if (!process.argv.slice(2).length) {
//     commander.help();
// }

// 提示声
var beep = require('beepbeep'); // 如果要更复杂的播放声音 https://github.com/Marak/play.js
var notifier = require('node-notifier');

// cron 定时任务
var schedule = require('node-schedule');

var crawlerUtil = require('./crawler-util.js');

/**
 * 在请求发出去之前, 做一些通用的默认设置
 * 例如设置统一的超时时间, 统一的 User-Agent 头
 */
function buildAgent(agent) {
    agent.set('User-Agent', 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.104 Safari/537.36 Core/1.53.1708.400 QQBrowser/9.5.9635.400');

    return agent;
}

// HTTP GET
function testGet() {
    var agent = request.get('http://httpbin.org/get');
    buildAgent(agent);

    return agent.query({
                    q: 1,
                    q2: '哈哈'
                })
                .then(function(result) {
                    return result.body;
                });
}
function testGetGb2312Html() {
    var agent = request.get('http://www.best73.com/Corp.html');
    buildAgent(agent);

    return agent.query({
                    q: 1,
                    q2: '哈哈'
                })
                .charset('gb2312')
                .then(crawlerUtil.showMeTheDollar)
                .then(function($) {
                    var title = '';
                    $('head title').each(function() {
                        var $this = $(this);
                        title = $this.text();
                    });

                    return title;
                });
}
// HTTP POST Content-Type: application/x-www-form-urlencoded
function testPostFormUrlEncoded() {
    var agent = request.post('http://httpbin.org/post');
    buildAgent(agent);

    return agent.type('form')
                .send({
                    q: 1,
                    q2: '哈哈'
                })
                .then(function(result) {
                    return result.body;
                });
}
// HTTP POST Content-Type: application/json
function testPostJson() {
    var agent = request.post('http://httpbin.org/post');
    buildAgent(agent);

    return agent.send({
                    q: 1,
                    q2: '哈哈'
                })
                .then(function(result) {
                    return result.body;
                });
}
// 测试通过 Promise 来控制延时任务
function testPromise() {
    return new Promise(function(resolve, reject) {
        testGet().then(function(json) {
            // console.log(json);
            setTimeout(function() {
                resolve(json);
            }, crawlerUtil.randomInt(2000, 3000));
        });
    });
}
// 测试序列任务
function testSequenceRequest() {
    return Array(10).fill(1).map((v, i) => i).reduce(function(promise, value) {
        return promise.then(function() {
            return testPostJson().then(function(t) {
                console.log(value, t);
                return t;
            });
        });
    }, Promise.resolve());
}
// 测试共享 Cookie
function testAgentCookie() {
    var agent = request.agent();
    agent.get('http://httpbin.org/cookies/set') // 第一个页面
         .query({
             name: 'crawler'
         })
         .then(function(result) {               // 第二个页面, cookie 会一直跟随着
             console.log(1, result.body.cookies);

             return agent.get('http://httpbin.org/cookies').then(function(result) {
                 console.log(2, result.body.cookies);
                 return result.body;
             });
         });
}
// 测试写入文件
function testWriteFile() {
    var dataFileName = '测试中文特殊:/\符号aaa.txt';
    dataFileName = sanitize(dataFileName);
    fs.writeFileSync(dataFileName, dataFileName);
    // fs.appendFileSync();
}
// 测试 Inquirer
function testInquirer() {
    var promps = [];
    promps.push({
        type: 'input',
        name: 'foo',
        message: '请输入',
        validate: function(input) {
            if(!input) {
                return '输入不能为空';
            }
            return true;
        }
    });
    promps.push({
        type: 'list',
        name: 'bar',
        message: '你想用要什么呢',
        choices: [{
            name: 'Text1',
            value: 't1'
        }, {
            name: 'Text2',
            value: 't2'
        }]
    });
    inquirer.prompt(promps).then(function(answers) {
        console.log(answers);
        console.log(chalk.red('我可以做一个命令行工具了...'));
    });
}
function testNotifySound() {
    beep(5, 1000);
}
function testNativeNotifier() {
    notifier.notify({
        title: '标题',
        message: '消息'
    });
}
function testDingdingRobot() {
    var webhook = 'https://oapi.dingtalk.com/robot/send?access_token=';
    var secret = 'SEC';

    return crawlerUtil.sendDingdingRobotMessage(webhook, secret, '消息').then(r => console.log(r));
}
function testDate() {
    var d = dayjs().startOf('month').add(1, 'day').set('year', 2018).format('YYYY-MM-DD HH:mm:ss');
    console.log(d);
}
function testCron() {
    var j = schedule.scheduleJob('*/5 * * * * *', function(){
        console.log(new Date().toLocaleTimeString(), '每 5 秒执行一次...');
        console.log('-----------------');
    });
}

// testNotifySound();
// testNativeNotifier();
// testDate();
// testDingdingRobot();
// testCron();
// testWriteFile();
// testInquirer();
// testAgentCookie();
// testGet().then(function(r) {console.log('testGet', r)});
// testGetGb2312Html().then(function(r) {console.log('testGetGb2312Html', r)});
// testPostFormUrlEncoded().then(function(r) {console.log('testPostFormUrlEncoded', r)});
// testPostJson().then(function(r) {console.log('testPostJson', r)});
// testPromise().then(function(r) {console.log('testPromise', r);});
// testSequenceRequest().then(function() {console.log('over');});